app/src/main/java/com/cdio/solitaire/controller/GameStateController.kt

```
@@ -0,0 +1,316 @@
+     * Assesses whether a stack move is legal.
+     */
+    private fun verifyMoveStack(move: Move): Boolean =
+        if (move.sourceStack == null || move.targetStack == null) {
+            Log.e("IllegalMoveError", "You need to specify a sourceStack and targetStack.")
+            false
+        } else tableauOrdering(move.sourceCard!!, move.targetStack!!)
+
+    /**
+     * Assesses whether a move from foundation is legal.
+     */
+    private fun verifyMoveFromFoundation(move: Move): Boolean {
+        val foundation = move.sourceStack
+        return if (foundation == null || move.targetStack == null) {
+            Log.e("IllegalMoveError", "You need to specify a sourceStack and targetStack.")
+            false
+        } else if (foundation.stackID in 8..11 && foundation.tail == move.sourceCard) {
+            tableauOrdering(move.sourceCard!!, move.targetStack!!)
+        } else {
+            false
+        }
+    }
+
+    /**
+     * Assesses whether a move from talon is legal.
+     */
+    private fun verifyMoveFromTalon(move: Move): Boolean =
+        if (gameState.talon.tail != move.sourceCard) {
+            false
+        } else if (move.targetStack!!.stackID in 8..11 && getFoundation(move.sourceCard!!.suit) == move.targetStack) {
+            true
+        } else if (move.targetStack!!.stackID in 1..7) {
+            verifyMoveStack(move)
+        } else false
+
+    /**
+     * Gets the foundation which has been used for the suit. If none exist, return the first empty foundation.
+     */
+    private fun getFoundation(suit: Suit): CardStack? {
+        var foundation: CardStack? = null
+        for (stack in gameState.foundations) {
+            if (stack.tail == null) {
+                foundation = stack
+                break
+            } else if (stack.tail!!.suit == suit) {
+                foundation = stack
+                break
+            } else {
+                Log.e("FoundationNotFound", "This shouldn't ever happen.")
```

Consider throwing an exception here instead of just logging the error, since a stack trace maybe could provide a lot of useful information.